### ISSUE TITLE
Phase 57.1: ExpertRouter — Token-to-Expert Routing & Gating Networks

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1082]
### Overview
The **ExpertRouter** implements intelligent token-to-expert assignment using learned gating networks. It supports multiple routing strategies including top-k gating (Shazeer et al.), Switch routing (Fedus et al.), and expert choice routing (Zhou et al.), enabling conditional computation where each token is processed by only a small subset of available experts.

### Architecture

```typescript
interface ExpertRouterConfig {
  numExperts: number;              // Total number of experts (e.g., 8, 16, 64, 128)
  topK: number;                    // Number of experts per token (1 for Switch, 2 for standard)
  routingStrategy: 'topk' | 'switch' | 'expert_choice' | 'hash' | 'base';
  gatingType: 'softmax' | 'sigmoid' | 'noisy_topk';
  capacityFactor: number;          // Expert capacity = (tokens/experts) * capacityFactor
  jitterNoise: number;             // Noise for load balancing during training (Shazeer et al.)
  auxiliaryLossWeight: number;     // Weight for auxiliary balancing loss
  routerZLossWeight: number;       // Weight for router z-loss (ST-MoE)
  expertDropout: number;           // Dropout rate for expert selection
  tokenDropPolicy: 'overflow' | 'random' | 'priority'; // How to handle capacity overflow
}

interface RoutingDecision {
  expertIndices: number[][];       // [batch*seq, topK] — selected expert indices per token
  routingWeights: number[][];      // [batch*seq, topK] — gating weights for combining expert outputs
  auxiliaryLoss: number;           // Load balancing auxiliary loss
  routerZLoss: number;             // Z-loss for routing stability
  expertLoads: number[];           // Tokens assigned to each expert
  droppedTokens: number;           // Count of tokens dropped due to capacity overflow
  routingEntropy: number;          // Entropy of routing distribution (diversity metric)
}

class ExpertRouter {
  // Core routing
  route(hiddenStates: Tensor, trainingMode: boolean): RoutingDecision;
  computeGatingScores(hiddenStates: Tensor): Tensor;  // [batch*seq, numExperts]
  applyTopKSelection(scores: Tensor, k: number): RoutingDecision;
  
  // Routing strategies
  noisyTopKGating(hiddenStates: Tensor): RoutingDecision;     // Shazeer et al.
  switchRouting(hiddenStates: Tensor): RoutingDecision;        // Fedus et al. top-1
  expertChoiceRouting(hiddenStates: Tensor): RoutingDecision;  // Zhou et al.
  hashRouting(tokenIds: Tensor): RoutingDecision;              // Roller et al.
  baseLayerRouting(hiddenStates: Tensor): RoutingDecision;     // Lewis et al.
  
  // Capacity management
  enforceCapacity(assignments: Tensor, capacityFactor: number): Tensor;
  handleOverflow(droppedTokens: Tensor, policy: string): Tensor;
  
  // Loss computation
  computeAuxiliaryLoss(gatingScores: Tensor, expertLoads: Tensor): number;
  computeRouterZLoss(logits: Tensor): number;  // ST-MoE stability loss
  computeImportanceLoss(gatingScores: Tensor): number;
  
  // Analysis
  getRoutingStatistics(): RoutingStats;
  visualizeExpertAssignment(): RoutingVisualization;
  computeRoutingEntropy(): number;
}
```

#### Noisy Top-K Gating (Shazeer et al. 2017)
```
H(x) = x · Wg                    # Linear projection to num_experts
noise = softplus(x · Wnoise)     # Learned noise for exploration
H(x) += noise * randn()          # Add noise during training
TopK = top_k(H(x), k)            # Select top-k experts
gates = softmax(TopK)             # Normalize selected gates
```

#### Switch Routing (Fedus et al. 2022)
```
router_logits = x · Wr            # Project to num_experts
routing_weights = softmax(router_logits)
expert_index = argmax(routing_weights)  # Top-1 selection
capacity = (n/E) * capacity_factor
# Route tokens respecting capacity constraint
```

#### Expert Choice Routing (Zhou et al. 2022)
```
S = softmax(X · We)^T            # [experts, tokens] score matrix
G, I = top_k(S, dim=1, k=C)     # Each expert selects top-C tokens
P = one_hot(I) * G               # Permutation with gating weights
Y = P^T · [E_i(P_i · X)]        # Aggregate expert outputs
```

### Dependencies
- Phase 56 (Interpretability) — routing visualization and expert analysis
- Phase 45 (Knowledge Distillation) — routing knowledge transfer

### Acceptance Criteria
- [ ] Top-k gating with configurable k (1, 2, 4, 8)
- [ ] Switch routing with capacity factor enforcement
- [ ] Expert choice routing implementation
- [ ] Auxiliary load balancing loss computation
- [ ] Router z-loss for training stability
- [ ] Token dropping with configurable policies
- [ ] Routing entropy monitoring
- [ ] Expert assignment visualization
- [ ] Unit tests with >95% coverage
- [ ] Benchmarks: <0.5ms routing overhead per layer for 128 experts

### References
1. Shazeer et al. "Outrageously Large Neural Networks" (2017)
2. Fedus et al. "Switch Transformers" (2022)
3. Zhou et al. "Mixture-of-Experts with Expert Choice Routing" (2022)
4. Roller et al. "Hash Layers for Large Sparse Models" (2021)
5. Lewis et al. "BASE Layers" (2021)
6. Zoph et al. "ST-MoE: Designing Stable and Transferable Sparse Expert Models" (2022)
